### HW2 
### Qiyun (Jenny) Ni
### Computing for Finance



# Question 1

In [1]:
from bs4 import BeautifulSoup
import urllib.request


def is_contain_covid(soup):
    '''
    Returns True if the input soup object contains "covid"
    text, and returns False otherwise. 
    '''
    text = soup.get_text()
    text = text.replace('\n', ' ')
    text = text.lower()
    return 'covid' in text


seed_url = "https://www.federalreserve.gov/newsevents/pressreleases.htm"


urls = [seed_url]    #queue of urls to crawl
seen = [seed_url]    #stack of urls seen so far
opened = []          #we keep track of seen urls so that we don't revisit them
contains_covid = []  #list of opened urls with containing "covid"

maxNumUrl = 15; #set the maximum number of urls to visit as 15

while len(urls) > 0 and len(contains_covid) < maxNumUrl:
    try:
        curr_url=urls.pop(0)
        req = urllib.request.Request(curr_url,headers={'User-Agent': 'Mozilla/5.0'})
        webpage = urllib.request.urlopen(req).read()
        soup = BeautifulSoup(webpage)
        opened.append(curr_url)
        if is_contain_covid(soup):
            print(curr_url)
            contains_covid.append(curr_url)

    except Exception as ex:
        print("Unable to access= "+curr_url)
        print(ex)
        continue    


    for tag in soup.find_all('a', href = True): #find tags with links
        childUrl = tag['href'] #extract just the link
        o_childurl = childUrl
        childUrl = urllib.parse.urljoin(seed_url, childUrl)

        if seed_url[:-5] in childUrl and childUrl not in seen:
            urls.append(childUrl)
            seen.append(childUrl)





https://www.federalreserve.gov/newsevents/pressreleases/2021-press.htm
https://www.federalreserve.gov/newsevents/pressreleases/2020-press.htm
https://www.federalreserve.gov/newsevents/pressreleases/monetary20220615a.htm
https://www.federalreserve.gov/newsevents/pressreleases/monetary20220504a.htm
https://www.federalreserve.gov/newsevents/pressreleases/monetary20220126a.htm
https://www.federalreserve.gov/newsevents/pressreleases/other20220523a.htm
https://www.federalreserve.gov/newsevents/pressreleases/enforcement20220405a.htm
https://www.federalreserve.gov/newsevents/pressreleases/other20220225a.htm
https://www.federalreserve.gov/newsevents/pressreleases/bcreg20220214a.htm
https://www.federalreserve.gov/newsevents/pressreleases/other20220114a.htm
https://www.federalreserve.gov/newsevents/pressreleases/monetary20211215a.htm
https://www.federalreserve.gov/newsevents/pressreleases/monetary20211103a.htm
https://www.federalreserve.gov/newsevents/pressreleases/monetary20210922a.htm
https://w

The following are the 15 output links containing the word 'covid'.  
https://www.federalreserve.gov/newsevents/pressreleases/2021-press.htm  
https://www.federalreserve.gov/newsevents/pressreleases/2020-press.htm  
https://www.federalreserve.gov/newsevents/pressreleases/monetary20220615a.htm  
https://www.federalreserve.gov/newsevents/pressreleases/monetary20220504a.htm  
https://www.federalreserve.gov/newsevents/pressreleases/monetary20220126a.htm  
https://www.federalreserve.gov/newsevents/pressreleases/other20220523a.htm  
https://www.federalreserve.gov/newsevents/pressreleases/enforcement20220405a.htm  
https://www.federalreserve.gov/newsevents/pressreleases/other20220225a.htm  
https://www.federalreserve.gov/newsevents/pressreleases/bcreg20220214a.htm  
https://www.federalreserve.gov/newsevents/pressreleases/other20220114a.htm  
https://www.federalreserve.gov/newsevents/pressreleases/monetary20211215a.htm  
https://www.federalreserve.gov/newsevents/pressreleases/monetary20211103a.htm  
https://www.federalreserve.gov/newsevents/pressreleases/monetary20210922a.htm  
https://www.federalreserve.gov/newsevents/pressreleases/monetary20210616a.htm  
https://www.federalreserve.gov/newsevents/pressreleases/monetary20210428a.htm  
  


# Question 2

In [2]:
from bs4 import BeautifulSoup
import urllib.request


def is_contain_covid(soup):
    '''
    Returns True if the input soup object contains "covid"
    text, and returns False otherwise. 
    '''
    text = soup.get_text()
    text = text.replace('\n', ' ')
    text = text.lower()
    return 'charges' in text


seed_url = "https://www.sec.gov/news/pressreleases"

urls = [(seed_url,'seed url')]    #queue of url and text tuples to crawl
seen = [(seed_url,'seed url')]    #stack of url and text tuples seen so far
opened = []          #we keep track of seen url and text tuples so that we don't revisit them
contains_charges = []  #list of opened url and text tuples with containing "covid"

maxNumUrl = 20; #set the maximum number of urls to visit as 20

while len(urls) > 0 and len(contains_charges) < maxNumUrl:
    try:
        curr_url_tuple=urls.pop(0)
        tag_text = curr_url_tuple[1]
        curr_url = curr_url_tuple[0]
        req = urllib.request.Request(curr_url,headers={'User-Agent': 'Mozilla/5.0'})
        webpage = urllib.request.urlopen(req).read()
        soup = BeautifulSoup(webpage)
        opened.append(curr_url)
        if is_contain_covid(soup):
            print(curr_url)
            print(tag_text)
            contains_charges.append((curr_url, tag_text))

    except Exception as ex:
        print("Unable to access= "+curr_url)
        print(ex)
        continue    


    for tag in soup.find_all('a', href = True): #find tags with links
        childUrl = tag['href'] #extract just the link
        o_childurl = childUrl
        childUrl = urllib.parse.urljoin(seed_url, childUrl)
        tag_text = tag.text

        if seed_url[:-1] in childUrl and childUrl not in [tup[0] for tup in seen]:
            # We need to remove the last "s" in order to find 20 child links
            urls.append((childUrl,tag_text))
            seen.append((childUrl,tag_text))
            





https://www.sec.gov/news/pressreleases
seed url
https://www.sec.gov/news/pressreleases#main-content

      Skip to main content
    
https://www.sec.gov/news/pressreleases.rss
Press Releases
https://www.sec.gov/news/pressrelease/2017-64.html
Executives Charged With Manipulating Company's Accounting Systems to Steal Money
https://www.sec.gov/news/pressrelease/2017-63.html
SEC Charges Firms Involved in Layering, Manipulation Schemes
https://www.sec.gov/news/pressrelease/2017-62.html
SEC Charges Marijuana-Related Company and Executives With Touting Bogus Revenues
https://www.sec.gov/news/pressrelease/2017-60.html
SEC Charges Mexico-Based Homebuilder in $3.3 Billion Accounting Fraud 
https://www.sec.gov/news/pressrelease/2017-52.html
SEC Staff Issues Guidance Update and Investor Bulletin on Robo-Advisers
https://www.sec.gov/news/pressrelease/2017-44.html
SEC Charges Chinese Citizens Who Reaped Massive Profits From Insider Trading on Comcast-Dreamworks Acquisition
https://www.sec.gov/news/p

The following are the 20 links and the corresponding tag text containing "charges":  

https://www.sec.gov/news/pressreleases  
seed url  
https://www.sec.gov/news/pressreleases#main-content  

      Skip to main content  
    
https://www.sec.gov/news/pressreleases.rss  
Press Releases  
https://www.sec.gov/news/pressrelease/2017-64.html  
Executives Charged With Manipulating Company's Accounting Systems to Steal Money  
https://www.sec.gov/news/pressrelease/2017-63.html  
SEC Charges Firms Involved in Layering, Manipulation Schemes  
https://www.sec.gov/news/pressrelease/2017-62.html  
SEC Charges Marijuana-Related Company and Executives With Touting Bogus Revenues  
https://www.sec.gov/news/pressrelease/2017-60.html  
SEC Charges Mexico-Based Homebuilder in $3.3 Billion Accounting Fraud   
https://www.sec.gov/news/pressrelease/2017-52.html  
SEC Staff Issues Guidance Update and Investor Bulletin on Robo-Advisers  
https://www.sec.gov/news/pressrelease/2017-44.html  
SEC Charges Chinese Citizens Who Reaped Massive Profits From Insider Trading on Comcast-Dreamworks Acquisition  
https://www.sec.gov/news/pressrelease/2016-270.html  
Company Settles Charges in Whistleblower Retaliation Case  
https://www.sec.gov/news/pressrelease/2016-252.html  
PIMCO Settles Charges of Misleading Investors About ETF Performance  
https://www.sec.gov/news/pressrelease/2016-250.html  
SEC Charges Asset Management Fund and Manager   
https://www.sec.gov/news/pressrelease/2016-102.html  
Brokerage Firm Charged With Anti-Money Laundering Failures  
https://www.sec.gov/news/pressrelease/2016-49.html  
SEC Charges Oregon-Based Investment Group and Executives With Defrauding Investors  
https://www.sec.gov/news/pressrelease/2016-26.html  
SEC Names C. Dabney O’Riordan and Alka Patel as Associate Regional Directors in Los Angeles Office  
https://www.sec.gov/news/pressrelease/2015-216.html  
SEC to Hold Equity Market Structure Advisory Committee Meeting on October 27  
https://www.sec.gov/news/pressrelease/2015-195.html  
SEC Charges Florida-Based CPA with Fraud for Issuing Bogus Audit Opinions  
https://www.sec.gov/news/pressrelease/2015-189.html  
SEC Announces Fraud Charges in Cross-Border Scheme to Secretly Control and Manipulate Stock of Chinese Companies After Reverse Mergers  
https://www.sec.gov/news/pressrelease/2015-163.html  
SEC Charges 32 Defendants in Scheme to Trade on Hacked News Releases  
https://www.sec.gov/news/pressrelease/2015-62  
SEC Charges Oregon-Based Defense Contractor With FCPA Violations  